llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -265,6 +281,76 @@ void collectCompositeElementsInfoRecursive(
   }
 }
 
+/// Recursively iterates over a composite type in order to collect information
+/// about its scalar elements.
+void collectCompositeElementsDefaultValuesRecursive(
+    const Module &M, Constant *C, unsigned &Index, unsigned &Offset,
+    std::vector<char> &DefaultValues) {
+  Type *Ty = C->getType();
+  if (auto *ArrTy = dyn_cast<ArrayType>(Ty)) {
+    for (size_t I = 0; I < ArrTy->getNumElements(); ++I) {
+      Constant *El = cast<Constant>(C->getOperand(I));
+      // TODO: this is a spot for potential optimization: for arrays we could
+      // just make a single recursive call here and use it to populate
+      // DefaultValues in a loop.
+      collectCompositeElementsDefaultValuesRecursive(M, El, Index, Offset,
+                                                     DefaultValues);
+    }
+  } else if (auto *StructTy = dyn_cast<StructType>(Ty)) {
+    const StructLayout *SL = M.getDataLayout().getStructLayout(StructTy);
+    for (size_t I = 0, E = StructTy->getNumElements(); I < E; ++I) {
+      Constant *El = cast<Constant>(C->getOperand(I));
+      // When handling elements of a structure, we do not use manually
+      // calculated offsets (which are sum of sizes of all previously
+      // encountered elements), but instead rely on data provided for us by
+      // DataLayout, because the structure can be unpacked, i.e. padded in
+      // order to ensure particular alignment of its elements.
+      unsigned LocalOffset = Offset + SL->getElementOffset(I);
+
+      // If there was some alignment, fill the data between values with zeros.
+      while (LocalOffset != DefaultValues.size())
+        DefaultValues.push_back(0);
+
+      collectCompositeElementsDefaultValuesRecursive(M, El, Index, LocalOffset,
+                                                     DefaultValues);
+    }
+    // Update "global" offset according to the total size of a handled struct
+    // type.
+    Offset += SL->getSizeInBytes();
+  } else if (auto *VecTy = dyn_cast<FixedVectorType>(Ty)) {
+    for (size_t I = 0; I < VecTy->getNumElements(); ++I) {
+      Constant *El = cast<Constant>(C->getOperand(I));
+      // TODO: this is a spot for potential optimization: for vectors we could
+      // just make a single recursive call here and use it to populate
+      // DefaultValues in a loop.
+      collectCompositeElementsDefaultValuesRecursive(M, El, Index, Offset,
+                                                     DefaultValues);
+    }
+  } else { // Assume that we encountered some scalar element
+    int NumBytes = Ty->getScalarSizeInBits() / CHAR_BIT;
```

Comment:
We need to be careful with `bool` here, as it could be represented as `i1`, which will result in zero bytes copied into a buffer. See line 292 or so:
```c++
Desc.Size = Ty->getPrimitiveSizeInBits() / 8 +     
            (Ty->getPrimitiveSizeInBits() % 8 != 0)
```